src/v/ssx/include/ssx/single_sharded.h

```diff
@@ -0,0 +1,183 @@
+///         the service is stopped.
+template<typename Service>
+class maybe_service {
+    using underlying_t = std::optional<Service>;
+    underlying_t _underlying;
+
+public:
+    maybe_service() = default;
+
+    template<class... Args>
+    explicit maybe_service(bool start, Args&&... args) {
+        if (start) {
+            _underlying.emplace(std::forward<Args>(args)...);
+        }
+    }
+
+    ss::future<> stop() {
+        return _underlying ? _underlying->stop() : ss::make_ready_future();
+    }
+
+    // the ones below are only safe to be called on the shards where the service
+    // has started
+
+    constexpr const Service* operator->() const noexcept {
+        return &*_underlying;
+    }
+    constexpr Service* operator->() noexcept { return &*_underlying; }
+
+    constexpr const Service& operator*() const& noexcept {
+        return *_underlying;
+    }
+    constexpr Service& operator*() & noexcept { return *_underlying; }
+    constexpr const Service&& operator*() const&& noexcept {
+        return &*_underlying;
+    }
+    constexpr Service&& operator*() && noexcept { return &*_underlying; }
+};
+
+/// Template helper to run a real service service on one core and an small dummy
+/// on other cores.
+///
+/// \tparam Service a class to be instantiated on the core.  Must expose
+///         a \c stop() method that returns a \c future<>, to be called when
+///         the service is stopped.
+template<typename Service>
+class single_sharded : ss::sharded<maybe_service<Service>> {
+    using base = ss::sharded<maybe_service<Service>>;
+    // init with most insane value to maximize the chances to blow up with
+    // segfault/asan if used; it shouldn't be used though
+    ss::shard_id _shard{std::numeric_limits<ss::shard_id>::max()};
+
+public:
+    /// Constructs an empty \c single_sharded object.  No instances of the
+    /// service are created.
+    single_sharded() noexcept = default;
+    single_sharded(const single_sharded& other) = delete;
+    single_sharded& operator=(const single_sharded& other) = delete;
+    /// Sharded object with T that inherits from peering_sharded_service
+    /// cannot be moved safely, so disable move operations.
+    single_sharded(single_sharded&& other) = delete;
+    single_sharded& operator=(single_sharded&& other) = delete;
+    /// Destroys a \c single_sharded object.  Must not be in a started state.
+    ~single_sharded() = default;
+
+    /// Starts \c Service by constructing an instance on the specified logical
+    /// core with a copy of \c args passed to the constructor.
+    ///
+    /// \param shard Which shard to start on
+    /// \param args Arguments to be forwarded to \c Service constructor
+    /// \return a \ref seastar::future<> that becomes ready when the instance
+    /// has been constructed.
+    template<typename... Args>
+    ss::future<> start_on(ss::shard_id shard, Args&&... args) noexcept {
```

Comment:
Sounds useful, but subject to details, such as whether we'd like to include the already busy shard 0 in the random choice? It may depend on how heavy or numerous the randomly placed services are. Since we don't have a ready use case for it as of yet I'd rather skip it until we need it. If you don't mind?

Reply:
Actually I'll merge it, as another thing is waiting for this change. Ping me if you feel strong about your suggestion and I'll consider adding it later.

Reply:
yes of course, I didn't mean to block the PR with my comment, I was just thinking out loud for future use cases and improvements.